lib/galaxy/webapps/base/webapp.py

```diff
@@ -322,6 +323,13 @@ def __init__(
             # This is a web request, get or create session.
             assert session_cookie
             self._ensure_valid_session(session_cookie)
+
+        if isinstance(self.user, User) and (self.user.social_auth or self.user.custos_auth):
+            success = self.app.authnz_manager.refresh(self)
```

Comment:
I would assume that is too costly to do on every request. As a simple hack i'd maybe look at the galaxy_session update_time and refresh the token after n minutes (where you could make n configurable). When the token is refreshed you also bump the update_time.

Longer term I think it's probably a good idea to have a separate process that just refreshes tokens for logged-in users and logs out users if the refresh fails.

Reply:
refresh would not happen at each request, inside the call we check if the token is about to expire and if not, we return.
https://github.com/galaxyproject/galaxy/pull/15300/files#diff-52d2ab08a1b269dab28ef1bb2feb00aef8eac9a200797994f34d1868b7de51c1R171
So, real refresh does happen every n minutes and expiration time is encoded in token, so no need to do it configurable.
To avoid confusion here, I could rename `refresh` -> `refresh_if_needed` or something like that

Reply:
Nice, that's good to know, and yes, `refresh_if_needed` sounds good. I suppose you could still have many requests at once triggering the refresh, but I think solving this would require an app-wide user-specific lock mechanism (probably via the database), we can postpone that.